src/System.Private.CoreLib/shared/System/Globalization/ISOWeek.cs

```diff
@@ -0,0 +1,159 @@
+        {
+            return ToDateTime(year, MinWeek, DayOfWeek.Monday);
+        }
+
+        // The year parameter represents an ISO week-numbering year (also called ISO year informally).
+        // Each week's year is the Gregorian year in which the Thursday falls.
+        // The first week of the year, hence, always contains 4 January.
+        // ISO week year numbering therefore slightly deviates from the Gregorian for some days close to 1 January.
+        public static DateTime GetYearEnd(int year)
+        {
+            return ToDateTime(year, GetWeeksInYear(year), DayOfWeek.Sunday);
+        }
+
+        // From https://en.wikipedia.org/wiki/ISO_week_date#Weeks_per_year:
+        //
+        // The long years, with 53 weeks in them, can be described by any of the following equivalent definitions:
+        //
+        // - Any year starting on Thursday and any leap year starting on Wednesday.
+        // - Any year ending on Thursday and any leap year ending on Friday.
+        // - Years in which 1 January and 31 December (in common years) or either (in leap years) are Thursdays.
+        //
+        // All other week-numbering years are short years and have 52 weeks.
+        public static int GetWeeksInYear(int year)
+        {
+            if (year < MinYear || year > MaxYear)
+            {
+                throw new ArgumentOutOfRangeException(nameof(year), SR.ArgumentOutOfRange_Year);
+            }
+
+            int P(int y) => (y + (y / 4) - (y / 100) + (y / 400)) % 7;
+
+            if (P(year) == 4 || P(year - 1) == 3)
+            {
+                return WeeksInLongYear;
+            }
+
+            return WeeksInShortYear;
+        }
+
+        // From https://en.wikipedia.org/wiki/ISO_week_date#Calculating_a_date_given_the_year,_week_number_and_weekday:
+        //
+        // This method requires that one know the weekday of 4 January of the year in question.
+        // Add 3 to the number of this weekday, giving a correction to be used for dates within this year.
+        //
+        // Multiply the week number by 7, then add the weekday. From this sum subtract the correction for the year.
+        // The result is the ordinal date, which can be converted into a calendar date.
+        //
+        // If the ordinal date thus obtained is zero or negative, the date belongs to the previous calendar year.
+        // If greater than the number of days in the year, to the following year.
+        public static DateTime ToDateTime(int year, int week, DayOfWeek dayOfWeek)
+        {
+            if (year < MinYear || year > MaxYear)
+            {
+                throw new ArgumentOutOfRangeException(nameof(year), SR.ArgumentOutOfRange_Year);
+            }
+
+            if (week < MinWeek || week > MaxWeek)
+            {
+                throw new ArgumentOutOfRangeException(nameof(week), SR.ArgumentOutOfRange_Week_ISO);
+            }
+
+            if ((int)dayOfWeek < 0 || (int)dayOfWeek > 7)
```

Comment:
>if ((int)dayOfWeek < 0 || (int)dayOfWeek > 7) [](start = 12, length = 45)

could you please add a little comment telling why we allow 7 here? thanks

Reply:
Done!

Reply:
Cool. let's wait for the CI to be green and then we can merge the changes. would be nice if you can get the test PR meanwhile.